xtask/src/main.rs

```diff
@@ -0,0 +1,605 @@
+
+    if !added_targets.is_empty() {
+        println!("⏳ installing targets");
+
+        let mut args: Vec<&str> = vec!["target", "add"];
+        args.extend(added_targets.iter().map(|s| s.as_str()));
+        let status = Command::new("rustup").args(&args).status().unwrap();
+        if !status.success() {
+            // since installing targets is the first thing we do, hard panic is OK enough (user would notice at this point)
+            panic!("Error installing targets: {}", added_targets.join(" "));
+        }
+    }
+
+    Ok(added_targets)
+}
+
+fn uninstall_targets(targets: Vec<String>) {
+    if !targets.is_empty() {
+        println!("⏳ uninstalling targets");
+
+        let mut cmd_and_args: Vec<&str> = vec!["rustup", "target", "remove"];
+        cmd_and_args.extend(targets.iter().map(|s| s.as_str()));
+
+        // only print uninstall errors so the user can fix those manually if needed
+        match run_command::<&str>(&cmd_and_args, None, &[]) {
+            Ok(_) => {}
+            Err(e) => {
+                eprintln!("Error uninstalling targets {}: {}", targets.join(" "), e);
+            }
+        }
+    }
+}
+
+fn test_book(errors: &mut Vec<String>) {
+    println!("🧪 book");
+    do_test(
+        || run_command::<&str>(&["cargo", "clean"], None, &[]),
+        "book",
+        errors,
+    );
+
+    do_test(
+        || {
+            run_command::<&str>(
+                &["cargo", "build", "--features", "unstable-test"],
+                None,
+                &[],
+            )
+        },
+        "book",
+        errors,
+    );
+
+    do_test(
+        || {
+            run_command(
+                &[
+                    "mdbook",
+                    "test",
+                    "-L",
+                    "../target/debug",
+                    "-L",
+                    "../target/debug/deps",
+                ],
+                Some("book"),
+                &[],
+            )
+        },
+        "book",
+        errors,
+    );
+}
+
+fn test_lint(errors: &mut Vec<String>) {
+    println!("🧪 lint");
+    do_test(
+        || run_command::<&str>(&["cargo", "clean"], None, &[]),
+        "lint",
+        errors,
+    );
+    do_test(
+        || run_command::<&str>(&["cargo", "fmt", "--all", "--", "--check"], None, &[]),
+        "lint",
+        errors,
+    );
+
+    do_test(
+        || run_command::<&str>(&["cargo", "clippy", "--workspace"], None, &[]),
+        "lint",
+        errors,
+    );
+}
+
+fn test_host(deny_warnings: bool, errors: &mut Vec<String>) {
+    println!("🧪 host");
+
+    let env = if deny_warnings {
+        vec![("RUSTFLAGS", "--deny warnings")]
+    } else {
+        vec![]
+    };
+
+    do_test(
+        || run_command::<&str>(&["cargo", "check", "--workspace"], None, &env),
+        "host",
+        errors,
+    );
+
+    do_test(
+        || {
+            run_command::<&str>(
+                &["cargo", "check", "--all", "--features", "unstable-test"],
```

Comment:
`cargo --all` and `cargo --workspace` are alias, right? let's stick to one of them? (you used `--all` above).

Reply:
I've (hopefully! please double check!) copied all the commands from the old `ci.yml` verbatim